tests/conftest.py

```diff
@@ -83,6 +83,99 @@ def two_ports_skb(netns):
     yield netns
 
 
+@pytest.fixture
+def three_ns_nat(netns):
+    """Fixture that creates three netns connected through veth pairs, a VIP and DNATing
+    packets in the middle"""
+    ipr = IPRoute()
+
+    # Create the netns
+    ns0 = netns.add("ns0")
+    ns1 = netns.add("ns1")
+    ns2 = netns.add("ns2")
+
+    # Create the veth pairs
+    ipr.link("add", ifname="veth01", peer="veth10", kind="veth")
+    ipr.link("add", ifname="veth12", peer="veth21", kind="veth")
+
+    # Wait until links appear
+    peer, veth = ipr.poll(
+        ipr.link, "dump", timeout=5, ifname=lambda x: x in ("veth01", "veth10")
+    )
+    peer, veth = ipr.poll(
+        ipr.link, "dump", timeout=5, ifname=lambda x: x in ("veth12", "veth21")
+    )
+
+    # Move ifaces to their netns
+    ipr.link("set", ifname="veth01", net_ns_fd="ns0")
+    ipr.link("set", ifname="veth10", net_ns_fd="ns1")
+    ipr.link("set", ifname="veth12", net_ns_fd="ns1")
+    ipr.link("set", ifname="veth21", net_ns_fd="ns2")
+
+    # Setup first pair
+    ns0.link("set", ifname="veth01", state="up")
+    ns1.link("set", ifname="veth10", state="up")
+    ns0.addr(
+        "add",
+        index=ns0.link_lookup(ifname="veth01")[0],
+        address="10.0.42.2",
+        prefixlen=24,
+    )
+    ns1.addr(
+        "add",
+        index=ns1.link_lookup(ifname="veth10")[0],
+        address="10.0.42.1",
+        prefixlen=24,
+    )
+
+    # Setup second pair
+    ns1.link("set", ifname="veth12", state="up")
+    ns2.link("set", ifname="veth21", state="up")
+    ns1.addr(
+        "add",
+        index=ns1.link_lookup(ifname="veth12")[0],
+        address="10.0.43.1",
+        prefixlen=24,
+    )
+    ns2.addr(
+        "add",
+        index=ns2.link_lookup(ifname="veth21")[0],
+        address="10.0.43.2",
+        prefixlen=24,
+    )
+
+    # Setup a VIP on ns1/lo
+    ns1.link("set", ifname="lo", state="up")
+    ns1.addr(
+        "add",
+        index=ns1.link_lookup(ifname="lo")[0],
+        address="10.0.255.1",
+        prefixlen=32,
+    )
+
+    # Enable forwarding and DNAT in ns1
+    netns.run("ns1", "sysctl", "-w", "net.ipv4.conf.all.forwarding=1")
+    netns.run(
```

Comment:
not a blocking comment as this could be considered once we end up having an actual need, but I was wondering if it would make sense to split the fixture into multiple ones and have basic building blocks that could be reused. Like: 3ns + dnat + vip.

Reply:
> Or even `ns(3) + dnat(1) + vip(1)`. Could be, but only if we find it more maintainable and if we see a real value in factorizing this. It's often tricky and tests setups are not easily split into reusable parts in my experience (like, there's always some small particularities that makes it special).

yes, no need to split now, as there's no actual need. Probably the ns part will be used in multiple tests. About the others, I agree that are more susceptible to slightly different characterizations.

The rest LGTM, so I'm going to approve.

Reply:
cannot be approved for the cherry-picked commit. Let's just consider this approved by me :)

Reply:
Not really, I must have seen grayed-out things when they were not.